### PR TITLE
ARROW-3446: [R] Document mapping of Arrow <-> R types

### PR DESCRIPTION
**Pull request by nealrichardson**

### CODE REVIEW

File: r/src/array_to_vector.cpp [hunk 1]
```diff
@@ -418,6 +418,7 @@ class Converter_Struct : public Converter {
   std::vector<std::shared_ptr<Converter>> converters;
 };
 
+// Shouldn't this cast before dividing? Otherwise we're doing integer division
```
Comment (fsaintjacques) [inline r/src/array_to_vector.cpp:421]: Depends what you want, integer division is much faster than floating division.

File: r/vignettes/arrow.Rmd [hunk 2]
```diff
@@ -86,7 +88,73 @@ to other applications and services that use Arrow. One example is Spark: the
...
+| Arrow type        | R type                   |
+|-------------------|--------------------------|
+| boolean           | logical                  |
+| int8              | integer                  |
+| int16             | integer                  |
+| int32             | integer                  |
+| int64             | bit64::integer64         |
+| uint8             | integer                  |
+| uint16            | integer                  |
+| uint32            | double                   |
```
Comment (fsaintjacques) [inline r/vignettes/arrow.Rmd:131]: I'm curious about the double, is it a copy-paste error, or it really casts to double.

File: r/vignettes/arrow.Rmd [hunk 3]
```diff
@@ -86,7 +88,73 @@ to other applications and services that use Arrow. One example is Spark: the
...
+| uint64            | -                        |
+| float16           | -                   |
```
Comment (fsaintjacques) [inline r/vignettes/arrow.Rmd:133]: Align.

File: r/vignettes/arrow.Rmd [hunk 4]
```diff
@@ -86,7 +88,73 @@ to other applications and services that use Arrow. One example is Spark: the
...
+| binary            | -                        |
+| fixed_size_binary | -                        |
+| date32            | Date                     |
+| date64            | POSIXct                  |
+| time32            | hms::difftime            |
+| time64            | hms::difftime            |
+| timestamp         | POSIXct                  |
+| duration          | -                        |
+| decimal           | double                   |
+| dictionary        | factor                   |
```
Comment (fsaintjacques) [inline r/vignettes/arrow.Rmd:146]: The factor is only for primitive type?

File: r/src/array_to_vector.cpp (hunk 1 above)
Comment (nealrichardson) [reply r/src/array_to_vector.cpp:421]: I would think we'd be more interested in fidelity of the data transformation than raw speed. Assuming I understand what the operators are doing, isn't `float(1234 / 1000)` a lot different from `float(1234) / 1000`?

File: r/vignettes/arrow.Rmd (hunk 2 above)
Comment (nealrichardson) [reply r/vignettes/arrow.Rmd:131]: It's "real" ;) https://github.com/apache/arrow/blob/master/r/src/array_to_vector.cpp#L717-L720

File: r/vignettes/arrow.Rmd (hunk 4 above)
Comment (nealrichardson) [reply r/vignettes/arrow.Rmd:146]: I don't understand the question, but I should make a footnote that dictionary values are coerced to string in order to be read as R factor (JIRA outstanding for making that configurable)

File: r/src/array_to_vector.cpp (hunk 1 above)
Comment (fsaintjacques) [reply r/src/array_to_vector.cpp:421]: Correct, you lose the fractional part, but yes we can revise that. I have no preference, it's probably better to preserve precision.

File: r/vignettes/arrow.Rmd (hunk 2 above)
Comment (fsaintjacques) [reply r/vignettes/arrow.Rmd:131]: @romainfrancois, why not using the same as int64_t?

File: r/vignettes/arrow.Rmd (hunk 4 above)
Comment (fsaintjacques) [reply r/vignettes/arrow.Rmd:146]: Yes, I meant as a footnote that we only support some sub-types.

File: r/vignettes/arrow.Rmd (hunk 2 above)
Comment (nealrichardson) [reply r/vignettes/arrow.Rmd:131]: Perhaps we could. The `bit64::integer64` class we use for int64 only supports signed integers, so `uint32` could fit in that, though `uint64` could not (cf. https://jira.apache.org/jira/browse/ARROW-9031).

I would also like to [pull int64, uint32, and uint64 into an R integer](https://jira.apache.org/jira/browse/ARROW-9083) if they are in bounds (32-bit signed).

In any case, this PR is merely documenting the status quo, so if we want to change uint32 behavior, let's open a new JIRA.

Comment (nealrichardson) [reply r/vignettes/arrow.Rmd:131]: Expanded https://jira.apache.org/jira/browse/ARROW-9083 to cover uint{32,64}. I'm going to proceed with this PR as is and will happily revise the docs if we change the implementation. 

Thanks for the discussion--shining a light on things like this is a big reason I picked up the documentation task.

**Review by fsaintjacques — APPROVED** [review]

File: r/src/array_to_vector.cpp (hunk 1 above)
Comment (nealrichardson) [reply r/src/array_to_vector.cpp]: I made the change and tests pass, so let's go with it.

**Pull request closed without merging**